### ISSUE TITLE
C++ ABI problems on macOS / with LLVM

### ISSUE THREAD

**Issue by eschnett** [JuliaPackaging/Yggdrasil#3223]
I am trying to build `openPMD_api` with Julia wrappers generated via `libcxxwrap_julia` for macOS (`x86_64-apple-darwin`). I receive linker errors of the kind

```
[13:16:03] Undefined symbols for architecture x86_64:
[13:16:03]   "jlcxx::julia_type(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
```
This symbol should be defined in the library `libcxxwrap_julia.dylib`. Unfortunately, this library defines a slightly different symbol:
```
jlcxx::julia_type(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
```

The difference, of course, are the inline namespace names `__cxx11` (GCC's libstdc++) vs. `__1` (LLVM's libc++).

I don't quite understand what is going on, but I have a few observations:
- BinaryBuilder looks for `libstdc++` when trying to detect the C++ ABI. It does not look for `libc++`, i.e. it wouldn't detect ABI incompatibilities coming from mixing up `libstdc++` and `libc++`.

... and while writing this report, I found the likely cause:

- The library `libcxxwrap_julia.dylib` the cmake option `-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN}`. I assume this uses LLVM on macOS.
- `openPMD_api` uses `-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN%.*}_gcc.cmake`, which presumably uses GCC on macOS.

If that is indeed the cause, then I assume BinaryBuilder should detect and handle a third kind of C++ ABI (LLVM ABI version 1). Note that `libc++` internally already offers an LLVM ABI version 2, so this might be used by some packages in the future. This version uses `__2` as inline namespace name; see https://github.com/llvm/llvm-project/blob/43fadefb0e77c56de7637c391cf98cf709b27095/libcxx/include/__config#L146 .

I observed a while back that macOS didn't seem to make a distinction between the C++03 and C++11 ABIs. I think the reason is the one above – on macOS, the default compilers is LLVM with ABI version 1, and the previous default compiler (a very old version of GCC) is not used in practice.

**Comment by giordano** [comment]
> BinaryBuilder looks for libstdc++ when trying to detect the C++ ABI. It does not look for libc++, i.e. it wouldn't detect ABI incompatibilities coming from mixing up libstdc++ and libc++.

The dual C++ string ABI is a GCC issue: https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html.  `libc++` is the LLVM C++ standard library, it doesn't have such a thing.

I suspect this may be relevant:
https://github.com/JuliaPackaging/Yggdrasil/blob/78ac9208dd89826716c5f2b283062d7a814e6b03/O/openPMD_api/build_tarballs.jl#L44-L45
Skipping macOS here was wrong, the idea that we had to skip it was influenced by a bug in BinaryBuilder fixed by https://github.com/JuliaPackaging/BinaryBuilder.jl/pull/1064.  It may be worth rebuilding `openPMD_api` with
```julia
platforms = expand_cxxstring_abis(platforms)
```
and trying again.

**Comment by eschnett** [comment]
I did that; it doesn't change things.

I believe the problem is that I'm building openPMD_api with GCC (and linked against libstdc++), while its dependency libcxxwrap_julia is built with LLVM (and linked against libc++). These two C++ standard libraries seem ABI incompatible.

**Comment by giordano** [comment]
Pinging @barche and @fingolfin to see if they have some other ideas.

Do you happen to remember why we used GCC for BSD platforms for openPMD_api?

**Comment by eschnett** [comment]
openPMD_api uses `std::variant` and `std::uncaught_exceptions` if C++17 is present. Unfortunately, Apple's implementations of these require at least macOS 10.14. Using these on earlier versions is explicitly disabled in Apple's libc++. BinaryBuilder uses darwin14, which is too old. I think darwin18 would work.

In other words, `-std=c++17` on darwin14 does not give you a full C++17.

My attempted work-around is now to change the code to avoid these features, and then to build with Clang. One could argue that openPMD_api should not use C++17 features without detecting whether these features are actually supported.

My long-term hope would be to make newer versions of Darwin available with BinaryBuilder.

**Comment by barche** [comment]
> My long-term hope would be to make newer versions of Darwin available with BinaryBuilder.

It's already possible to use newer Darwin SDKs, as done in e.g. the Qt packages: https://github.com/JuliaPackaging/Yggdrasil/blob/master/Q/Qt5Base/build_tarballs.jl

If supported in the dependency specs, maybe providing both GCC and LLVM versions of libcxxwrap could also be an option.

**Comment by eschnett** [comment]
Thanks for the pointer to building with macOS 10.14.

libcxxwrap depends on Julia. Building Julia on macOS with GCC is probably very involved.

**Comment by barche** [comment]
> libcxxwrap depends on Julia. Building Julia on macOS with GCC is probably very involved

We could cheat there I think, since the interface between libjulia and libcxxwrap-julia is only C, not C++, so keeping the LLVM Julia will probably just work. Anyway, I would only go for this solution if there is no alternative.

**Comment by Gnimuc** [comment]
What's the `-stdlib=` compiler flag used when compiling `openPMD_api`? The macOS default should be `-stdlib=libc++` and this flag is macOS-only. 

I hit a similar issue when I was trying to programmatically compile some code with BB2's system headers on macOS. The code was going to be linked to libraries shipped by BB2 artifacts, but I mistakenly used those stdlibc++ headers.

https://libcxx.llvm.org//UsingLibcxx.html#using-a-different-version-of-the-c-standard

**Comment by eschnett** [comment]
I have now switched to building all Yggdrasil packages with clang when on macOS, which is also the default choice of BinaryBuilder.

Some background: Originally, I encountered problems with using clang for some packages, and I thought switching to GCC would be easy. I was then not aware of the ABI incompatibility between GCC and Clang (i.e. between libstdc++ and libc++). It doesn't help that BinaryBuilder doesn't detect this ABI difference; instead, it uses the same ABI tag for libc++ as for the libstdc++/C++11 mode, so one has to manually ensure not to mix these.

As you mention, setting GCC to use the system libc++ might work on macOS. I did not attempt to do this. GCC, by default, uses its own libstdc++ on all platforms, including on macOS.

**Comment by Gnimuc** [comment]
Yeah. To work better with BB2, we should compromise BB2's environment. 

Recently, I just build everything with BB2's system headers instead of those on my local machine. The idea is to pull those BB2 shards and setup includes with a combination of `-isystem`&`-nostdlib/-nostdlib++`&`-nostdinc/-nostdinc++`: 

https://github.com/Gnimuc/ClangComplier.jl/blob/master/src/platform/system.jl#L42-L77 

https://github.com/Gnimuc/ClangComplier.jl/blob/master/src/ClangCompiler.jl#L20-L25
